std/evmprecompiles/15-blspairing.go

```diff
@@ -0,0 +1,114 @@
+package evmprecompiles
+
+import (
+	"fmt"
+
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/std/algebra/emulated/sw_bls12381"
+)
+
+// ECPairBLS implements [BLS12_PAIRING_CHECK] precompile contract at address 0x0f.
+//
+// To have a fixed-circuit regardless of the number of inputs, we need 2 fixed circuits:
+//   - MillerLoopAndMul:
+//     A Miller loop of fixed size 1 followed by a multiplication in 𝔽p¹².
+//   - MillerLoopAndFinalExpCheck:
+//     A Miller loop of fixed size 1 followed by a multiplication in 𝔽p¹², and
+//     a check that the result lies in the same equivalence class as the
+//     reduced pairing purported to be 1. This check replaces the final
+//     exponentiation step in-circuit and follows Section 4 of [On Proving
+//     Pairings] paper by A. Novakovic and L. Eagen.
+//
+// N.B.: This is a sub-optimal routine but defines a fixed circuit regardless
+// of the number of inputs.  We can extend this routine to handle a 2-by-2
+// logic but we prefer a minimal number of circuits (2).
+//
+// See the methods [ECPairMillerLoopAndMul] and [ECPairMillerLoopAndFinalExpCheck] for the fixed circuits.
+// See the methods [ECPairBLSIsOnG1] and [ECPairBLSIsOnG2] for the check that Pᵢ and Qᵢ are on G1 and resp. G2.
+//
+// [BLS12_PAIRING_CHECK]: https://eips.ethereum.org/EIPS/eip-2537
+// [On Proving Pairings]: https://eprint.iacr.org/2024/640.pdf
+func ECPairBLS(api frontend.API, P []*sw_bls12381.G1Affine, Q []*sw_bls12381.G2Affine) {
+	if len(P) != len(Q) {
+		panic("P and Q length mismatch")
+	}
+	if len(P) < 2 {
+		panic("invalid multipairing size bound")
+	}
+	n := len(P)
+	pair, err := sw_bls12381.NewPairing(api)
+	if err != nil {
+		panic(err)
+	}
+	for i := 0; i < n; i++ {
+		// 1- Check that Pᵢ are on G1
```

Comment:
I would add explicitly the comment why in this case we do it here, not in the zkevm. Then imo better to review later when comparing to bn254 pairing precompile